File: C++/remove-nth-node-from-end-of-list.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* removeNthFromEnd(ListNode* head, int n) {
        int count = countNodes(head);
        int target = count - n;

        if (target <= 0) {
            return head->next;
        }

        ListNode* node = head;

        for (int i = 0; i < target - 1; i++) {
            node = node->next;
        }
        
        node->next = node->next->next;
        
        return head;        
    }
    
    int countNodes(ListNode* node) {
        int count = 0;
        for (; node; node = node->next) {
            count++;
        }
        return count;
    }
};

File: Python3/remove-nth-node-from-end-of-list.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def removeNthFromEnd(self, head: Optional[ListNode], n: int) -> Optional[ListNode]:
        node = head
        count = 0
        while node:
            count += 1
            node = node.next
        removeN = count - n
        if removeN == 0:
            return head.next
        node = head
        while removeN > 1:
            removeN -= 1
            node = node.next
        node.next = node.next.next
        return head

        